### PR TITLE
test/ integration tests for update impersonation request

### LINKED ISSUE #2200: Impersonation Feature For RDS priviledged users(super_user) #2
### Feature/Issue Description
User impersonation feature needs to be implemented which allows a privileged user (i.e., super-user) to impersonate a regular user, enabling the super-user to view the regular user’s view, such as tasks and progress, without logging in as the regular user.

### User Stories
 1) Privileged User creates an impersonation request for a normal user
 2) The normal user can approve/reject this request
 3) Privileged user can start and perform impersonation once for an approved request for a fixed decided time
 4) Privileged user gets a read-only access when impersonating
 5) Privileged user can end the impersonation at any time within the time limit

### Expected Behavior

- To have a request-based mechanism so that a privileged user can create a request when he wants to impersonate a normal user,
- On receiving a request from a privileged user, a normal user can approve/reject it, if approved, he will give an access to impersonate him for a fixed amount of time,
- Restrict the impersonation access to the privileged user so that he only gets a read-only view when impersonating,
- At any point, the privileged user can stop this impersonation session and go back to normal

### Current Behavior
We currently have no existing way/feature for impersonating a user,

 PRD - [LINK](https://docs.google.com/document/d/1KI7zp81vevaIX3nDNgTsCHDY1iWmad-XFaVgEi3nH2Q/edit?tab=t.0) 

### Screenshots
NA

### Reproducibility

 - [ ] This issue is reproducible
 - [x] This issue is not reproducible

### Steps to Reproduce
NA

### Severity/Priority
 - [ ] Critical
 - [ ] High
 - [x] Medium
 - [ ] Low

### Additional Information
[Initial PRD](https://docs.google.com/document/d/1TJ3-AfeiVajmYMxXW0u79rrOOHt81lZuntX2v9dr638/edit)

### Checklist

- [x] I have read and followed the project's code of conduct.
- [x] I have searched for similar issues before creating this one.
- [x] I have provided all the necessary information to understand and reproduce the issue.
- [x] I am willing to contribute to the resolution of this issue.

### PR DESCRIPTION
**Pull request by Suvidh-kaushik**
Date: 21/06/2025

Developer Name: Suvidh Kaushik

---

## Issue Ticket Number

- Closes #2200 

## Description

- This PR contains the integration tests required for the update Impersonation Requests feature.
-  **This PR will be merged after the feature PR #2446  is merged**
- **The import errors build fail is because our feature PR has not yet been merged**

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

N/A

</details>

## Test Coverage
<details>
<summary>Screenshots</summary>

Screenshot 2025-06-21 172648

Screenshot 2025-06-21 173145

Screenshot 2025-06-21 173502

Screenshot 2025-06-21 173054

Screenshot 2025-06-21 173104

</details>

## Additional Notes
- Design Doc - [LINK](https://docs.google.com/document/d/1Qt0-knoFEWDl6Gdxr0W1xhPy87H-ixwj9Y55Q0bKMUo/edit?tab=t.0#heading=h.w1nbw3goorza)
- PRD - [LINK](https://docs.google.com/document/d/1KI7zp81vevaIX3nDNgTsCHDY1iWmad-XFaVgEi3nH2Q/edit?tab=t.0)

### What change is being made?
Add integration tests for the update impersonation request feature, including tests for various scenarios such as successful updates, unauthorized access, and request status validation.

### Why are these changes being made?
These changes enhance the test coverage of the impersonation request feature, ensuring the correct functionality and handling of edge cases such as non-existent requests, unauthorized updates, and error conditions, thereby reducing potential bugs and improving code reliability.

> Is this description stale? Ask me to generate a new description by commenting `/korbit-generate-pr-description`

### CODE REVIEW

File: test/integration/impersonationRequests.test.ts [hunk 1]
```diff
@@ -11,65 +11,83 @@ import addUser from "../utils/addUser";
...
   });
 
   afterEach(async () => {
-    sinon.restore();
     await cleanDb();
+    sinon.restore();
   });
 
-  describe("POST /impersonation/requests", () => {
+    describe("POST /impersonation/requests", () => {
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/integration/impersonationRequests.test.ts:90]: fixed

File: test/integration/impersonationRequests.test.ts [hunk 2]
```diff
@@ -264,4 +282,248 @@ describe("Impersonation Requests", () => {
...
+          if (err) return done(err);
+          try {
+            expect(res).to.have.status(401);
+            expect(res.body.error).to.equal("Unauthorized");
+            expect(res.body.message).to.equal("Unauthenticated User");
+            done();
+          } catch (e) {
+            done(e);
+          }
+        });
```
[the remark this answers is not in the document]
Comment (MayankBansal12) [reply test/integration/impersonationRequests.test.ts:433]: @Suvidh-kaushik please remove this duplicate test

File: test/integration/impersonationRequests.test.ts [hunk 3]
```diff
@@ -264,4 +282,248 @@ describe("Impersonation Requests", () => {
...
+        .set("cookie", `${cookieName}=${authToken}`)
+        .end(function (err, res) {
+          if (err) return done(err);
+          expect(res).to.have.status(403);
+          expect(res.body.error).to.equal("Forbidden");
+          expect(res.body.message).to.equal(UNAUTHORIZED_TO_UPDATE_REQUEST);
+          done();
+        });
+    });
+  });
```
Comment (MayankBansal12) [inline test/integration/impersonationRequests.test.ts:528]: nit: let's move this above to maintain consistency

Comment (Suvidh-kaushik) [reply test/integration/impersonationRequests.test.ts]: fixed

**Review by MayankBansal12 — APPROVED** [review]

File: test/integration/impersonationRequests.test.ts [hunk 4]
```diff
@@ -578,6 +589,232 @@ describe("Impersonation Requests", () => {
...
+        });
+    });
+
+    it("should return 404 and 'Route not found' message when dev is missing", function (done) {
+      chai
+        .request(app)
+        .patch(`/impersonation/requests/${impersonationRequest1.id}`)
+        .send({status:"APPROVED"})
+        .set("cookie", `${cookieName}=${authToken}`)
+        .end(function (err, res) {
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/integration/impersonationRequests.test.ts:648]: fixed

**Review by MayankBansal12 — APPROVED** [review]
